source/event_stream_rpc_server.c

```diff
@@ -0,0 +1,906 @@
+    aws_event_stream_rpc_server_message_flush_fn *flush_fn,
+    void *user_data) {
+    if (aws_event_stream_rpc_server_continuation_is_closed(continuation)) {
+        return aws_raise_error(AWS_ERROR_EVENT_STREAM_RPC_STREAM_CLOSED);
+    }
+
+    return s_send_protocol_message(
+        continuation->connection, continuation, message_args, continuation->stream_id, flush_fn, user_data);
+}
+
+/* just a convenience function for fetching message metadata from the event stream headers on a single iteration. */
+static int s_fetch_message_metadata(
+    struct aws_array_list *message_headers,
+    int32_t *stream_id,
+    int32_t *message_type,
+    int32_t *message_flags,
+    struct aws_byte_buf *operation_name) {
+    size_t length = aws_array_list_length(message_headers);
+    size_t required_fields_found = 0;
+    size_t optional_fields_found = 0;
+
+    for (size_t i = 0; i < length; ++i) {
+        struct aws_event_stream_header_value_pair *header = NULL;
+        aws_array_list_get_at_ptr(message_headers, (void **)&header, i);
+        struct aws_byte_buf name_buf = aws_event_stream_header_name(header);
+
+        /* check type first since that's cheaper than a string compare */
+        if (header->header_value_type == AWS_EVENT_STREAM_HEADER_INT32) {
+
+            struct aws_byte_buf stream_id_field = aws_byte_buf_from_array(
+                aws_event_stream_rpc_stream_id_name.ptr, aws_event_stream_rpc_stream_id_name.len);
+            if (aws_byte_buf_eq_ignore_case(&name_buf, &stream_id_field)) {
+                *stream_id = aws_event_stream_header_value_as_int32(header);
+                goto found;
+            }
+
+            struct aws_byte_buf message_type_field = aws_byte_buf_from_array(
+                aws_event_stream_rpc_message_type_name.ptr, aws_event_stream_rpc_message_type_name.len);
+            if (aws_byte_buf_eq_ignore_case(&name_buf, &message_type_field)) {
+                *message_type = aws_event_stream_header_value_as_int32(header);
+                goto found;
+            }
+
+            struct aws_byte_buf message_flags_field = aws_byte_buf_from_array(
+                aws_event_stream_rpc_message_flags_name.ptr, aws_event_stream_rpc_message_flags_name.len);
+            if (aws_byte_buf_eq_ignore_case(&name_buf, &message_flags_field)) {
+                *message_flags = aws_event_stream_header_value_as_int32(header);
+                goto found;
+            }
+        }
+
+        if (header->header_value_type == AWS_EVENT_STREAM_HEADER_STRING) {
+            struct aws_byte_buf operation_field = aws_byte_buf_from_array(
+                aws_event_stream_rpc_operation_name.ptr, aws_event_stream_rpc_operation_name.len);
+
+            if (aws_byte_buf_eq_ignore_case(&name_buf, &operation_field)) {
+                *operation_name = aws_event_stream_header_value_as_string(header);
+                goto optional_found;
+            }
+        }
+
+    found:
+        required_fields_found += 1;
```

Comment:
This is going to fall through every time and increment ```optional_fields_found```, I don't think that's the intent here?

Reply:
yikes! good catch